File: frontend/src/types/serial.ts
export type SerialConnectionState =
  | 'DISCONNECTED'
  | 'MOCK_CONNECTED'
  | 'PORT_OPEN_NO_TELEMETRY'
  | 'READONLY_CONNECTED_UNVERIFIED'
  | 'PICO_READONLY_VERIFIED'
  | 'MOCK_READONLY_CONNECTED'
  | 'FAULT'
export type SerialDirection = 'tx' | 'rx' | 'system'
export type SerialLogKind = 'tx' | 'rx' | 'ack' | 'nack' | 'error' | 'timeout' | 'status'

export interface SerialStatus {
  connection_state: SerialConnectionState
  transport_mode: string
  transport_source: string
  protocol_mode: string
  real_serial_enabled: boolean
  real_serial_readonly: boolean
  readonly: boolean
  telemetry_received: boolean
  pico_verified: boolean
  physical_command_enabled: boolean
  last_tx: Record<string, unknown> | null
  last_rx: Record<string, unknown> | null
  pending_ack_count: number
  command_queue_depth: number
  last_command_age_ms: number | null
  last_command_kind: string | null
  last_command_raw: string | null
  last_command_ack_state: string
  last_command_rtt_ms: number | null
  last_command_error: string | null
  magazine_capacity: number
  magazine_remaining: number
  magazine_empty: boolean
  acknowledged_shot_count: number
  magazine_reload_count: number
  magazine_updated_at: number | null
  heartbeat_age_ms: number | null
  ack_timeout_ms: number
  heartbeat_timeout_ms: number
  last_error: string | null
}

export interface SerialLogEntry {
  id: number
  ts: number
  direction: SerialDirection
  kind: SerialLogKind
  message: Record<string, unknown>
  raw: string | null
  error: string | null
}

export interface SerialCommandResult {
  accepted: boolean
  reason: string
  status: SerialStatus
  log_entry: SerialLogEntry | null
  no_physical_command_generated: boolean
}

File: backend/app/schemas/serial.py
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class SerialConnectionState(StrEnum):
    DISCONNECTED = "DISCONNECTED"
    MOCK_CONNECTED = "MOCK_CONNECTED"
    PORT_OPEN_NO_TELEMETRY = "PORT_OPEN_NO_TELEMETRY"
    READONLY_CONNECTED_UNVERIFIED = "READONLY_CONNECTED_UNVERIFIED"
    PICO_READONLY_VERIFIED = "PICO_READONLY_VERIFIED"
    MOCK_READONLY_CONNECTED = "MOCK_READONLY_CONNECTED"
    FAULT = "FAULT"


class SerialDirection(StrEnum):
    TX = "tx"
    RX = "rx"
    SYSTEM = "system"


class SerialLogKind(StrEnum):
    TX = "tx"
    RX = "rx"
    ACK = "ack"
    NACK = "nack"
    ERROR = "error"
    TIMEOUT = "timeout"
    STATUS = "status"


class SerialStatus(BaseModel):
    connection_state: SerialConnectionState
    transport_mode: str
    transport_source: str = "mock"
    protocol_mode: str
    real_serial_enabled: bool
    real_serial_readonly: bool = True
    readonly: bool = False
    telemetry_received: bool = False
    pico_verified: bool = False
    physical_command_enabled: bool = False
    last_tx: dict[str, Any] | None = None
    last_rx: dict[str, Any] | None = None
    pending_ack_count: int = 0
    command_queue_depth: int = 0
    last_command_age_ms: int | None = None
    last_command_kind: str | None = None
    last_command_raw: str | None = None
    last_command_ack_state: str = "unknown"
    last_command_rtt_ms: int | None = None
    last_command_error: str | None = None
    magazine_capacity: int = 8
    magazine_remaining: int = 8
    magazine_empty: bool = False
    acknowledged_shot_count: int = 0
    magazine_reload_count: int = 0
    magazine_updated_at: float | None = None
    heartbeat_age_ms: int | None = None
    ack_timeout_ms: int
    heartbeat_timeout_ms: int
    last_error: str | None = None


class SerialLogEntry(BaseModel):
    id: int
    ts: float
    direction: SerialDirection
    kind: SerialLogKind
    message: dict[str, Any]
    raw: str | None = None
    error: str | None = None


class SerialSendJsonRequest(BaseModel):
    message: dict[str, Any]


class SerialSimulateRxRequest(BaseModel):
    message: dict[str, Any]


class SerialCommandResult(BaseModel):
    accepted: bool
    reason: str
    status: SerialStatus
    log_entry: SerialLogEntry | None = None
    no_physical_command_generated: bool = True
